src/RegistryHandler.ts

```diff
@@ -1,79 +1,74 @@
 	}
 
-	public remove(registry: Registry): boolean {
-		return this._registries.some((registryWrapper, i) => {
-			if (registryWrapper.registry === registry) {
-				registry.destroy();
-				this._registries.splice(i, 1);
-				return true;
-			}
-			return false;
-		});
+	public set base(baseRegistry: Registry) {
+		this._registryWidgetLabelMap.delete(this._baseRegistry);
+		this._registryInjectorLabelMap.delete(this._baseRegistry);
+		this._baseRegistry = baseRegistry;
 	}
 
-	public replace(original: Registry, replacement: Registry): boolean {
-		return this._registries.some((registryWrapper, i) => {
-			if (registryWrapper.registry === original) {
-				original.destroy();
-				registryWrapper.registry = replacement;
-				return true;
-			}
-			return false;
-		});
+	public define(label: RegistryLabel, widget: Constructor<WidgetBaseInterface>): void {
+		this._registry.define(label, widget);
 	}
 
-	public get defaultRegistry(): Registry | undefined {
-		if (this._registries.length) {
-			return this._registries[this._registries.length - 1].registry;
-		}
+	public defineInjector(label: RegistryLabel, injector: Injector): void {
+		this._registry.defineInjector(label, injector);
 	}
 
-	public has(widgetLabel: RegistryLabel): boolean {
-		return this._registries.some((registryWrapper) => {
-			return registryWrapper.registry.has(widgetLabel);
-		});
+	public has(label: RegistryLabel): boolean {
+		return this._registry.has(label) || this._baseRegistry.has(label);
 	}
 
-	public getInjector<T extends Injector>(label: RegistryLabel): T | null {
-		for (let i = 0; i < this._registries.length; i++) {
-			const registryWrapper = this._registries[i];
-			return registryWrapper.registry.getInjector<T>(label);
-		}
-		return null;
+	public hasInjector(label: RegistryLabel): boolean {
+		return this._registry.hasInjector(label) || this._baseRegistry.hasInjector(label);
 	}
 
-	public get<T extends WidgetBaseInterface = WidgetBaseInterface>(widgetLabel: RegistryLabel): Constructor<T> | null {
-		for (let i = 0; i < this._registries.length; i++) {
-			const registryWrapper = this._registries[i];
-			const item = registryWrapper.registry.get<T>(widgetLabel);
+	public get<T extends WidgetBaseInterface = WidgetBaseInterface>(label: RegistryLabel, globalPrecedence: boolean = false): Constructor<T> | null {
+		return this._get(label, globalPrecedence, 'get', this._registryWidgetLabelMap);
+	}
+
+	public getInjector(label: RegistryLabel, globalPrecedence: boolean = false): Injector | null {
+		return this._get(label, globalPrecedence, 'getInjector', this._registryInjectorLabelMap);
+	}
+
+	private _get(label: RegistryLabel, globalPrecedence: boolean, getFunctionName: 'getInjector' | 'get', labelMap: Map<Registry, RegistryLabel[]>): any {
+		const registries = globalPrecedence ? [ this._baseRegistry, this._registry ] : [ this._registry, this._baseRegistry ];
+		for (let i = 0; i < registries.length; i++) {
+			const registry: any = registries[i];
```

Comment:
No other decent type than `any`?

Reply:
It's so we can use an index string function name that is passed. 

If the `getFunctionName` (which typing I realise is a bad name), is restricted to just `'get'` or `'getInjector'` then using `registry[getFunctionName]('')` works but not if is it's typed as `'get' | 'getInjector'`, any ideas?

```ts
private _get(label: RegistryLabel, globalPrecedence: boolean, getFunctionName: 'getInjector' | 'get', labelMap: Map<Registry, RegistryLabel[]>): any {
```

Reply:
Hmmm... [Recent versions of TypeScript](https://www.typescriptlang.org/play/#src=class%20Registry%20%7B%0D%0A%20%20%20%20getInjector()%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20'foo'%3B%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20get()%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20'bar'%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Afunction%20getThings(method%3A%20'get'%20%7C%20'getInjector'%2C%20registry%3A%20Registry)%20%7B%0D%0A%20%20%20%20return%20registry%5Bmethod%5D()%3B%0D%0A%7D%0D%0A%0D%0AgetThings('get'%2C%20new%20Registry())%3B%0D%0A) allow you to access properties by method name just fine, if the variable is a string literal, unless I am missing something here.

Reply:
It was fine is I typed the argument is just `'get'` or just `'getInjector'` but not when I typed it as `'getInjector' | 'get'`...

Reply:
Strange, the playground example types it as `'getInjector' | 'get'`... wonder what is wrong.